internal/config/ociartifact/ociartifact.go

```
@@ -24,54 +29,90 @@ func New() Impl {
 
-	storeOpts, err := storage.DefaultStoreOptions(false, 0)
+	name, err := libimage.NormalizeName(img)
 	if err != nil {
-		return nil, fmt.Errorf("get default storage options: %w", err)
+		return nil, fmt.Errorf("parse image name: %w", err)
 	}
 
-	store, err := storage.GetStore(storeOpts)
+	ref, err := docker.NewReference(name)
 	if err != nil {
-		return nil, fmt.Errorf("get container storage: %w", err)
+		return nil, fmt.Errorf("create docker reference: %w", err)
 	}
 
-	runtime, err := libimage.RuntimeFromStore(store, &libimage.RuntimeOptions{SystemContext: sys})
+	src, err := ref.NewImageSource(ctx, sys)
 	if err != nil {
-		return nil, fmt.Errorf("create libimage runtime: %w", err)
+		return nil, fmt.Errorf("build image source: %w", err)
 	}
 
-	images, err := runtime.Pull(ctx, ref, config.PullPolicyAlways, &libimage.PullOptions{})
+	manifestBytes, mimeType, err := src.GetManifest(ctx, nil)
 	if err != nil {
-		return nil, fmt.Errorf("pull OCI artifact: %w", err)
+		return nil, fmt.Errorf("get manifest: %w", err)
+	}
+
+	parsedManifest, err := manifest.FromBlob(manifestBytes, mimeType)
+	if err != nil {
+		return nil, fmt.Errorf("parse manifest: %w", err)
+	}
+	if enforceConfigMediaType != "" && parsedManifest.ConfigInfo().MediaType != enforceConfigMediaType {
+		return nil, fmt.Errorf(
+			"wrong config media type %q, requires %q",
+			parsedManifest.ConfigInfo().MediaType, enforceConfigMediaType,
+		)
 	}
-	image := images[0]
 
-	mountPath, err := image.Mount(ctx, nil, "")
+	layers := parsedManifest.LayerInfos()
+	if len(layers) < 1 {
+		return nil, errors.New("artifact needs at least one layer")
+	}
+	layer := layers[0]
+
+	bic := blobinfocache.DefaultCache(sys)
+	rc, _, err := src.GetBlob(ctx, layer.BlobInfo, bic)
+	if err != nil {
+		return nil, fmt.Errorf("get layer blob: %w", err)
+	}
+	defer rc.Close()
+
+	layerBytes, err := io.ReadAll(io.LimitReader(rc, maxArtifactSize))
+	if len(layerBytes) == maxArtifactSize {
```

When my image happens to be exactly 1 MiB, it will be a failure?

The `GetBlob()` function returns size if `Content-Size` header is present in the response (we assume remote server is well-behaved), so we could leverage this first to make a check and fail early (no need to actually read anything then), and if we got `-1`, then we could rely on this check here.

It would be nice to make this check a > 1 MiB, if at all possible.

Thoughts?

---

(Compare the implementation in `c/image/internal/iolimits`.)

---

Done, fixed.